# Introduction to SQL

Here you can access the `books` table used in the course.

--- 
_Note: When using sample integrations such as those that contain course data, you have read-only access. You can run queries, but cannot make any changes such as adding, deleting, or modifying the data (e.g., creating tables, views, etc.)._

## Course goals
1. Understand databases and their structure (Chapter 1)
2. Extract information from databases using SQL (Chapter 2)

### Chapter 1 - Relational Databases

**Database**
- stores data
- information are housed in an object called **tables** , with data organized into _columns_ and _rows_

**Relational Databases**
- relationships between tables of data inside the database.
![image](image.png)

**Database advantages**
-  more storage than spreasheet applications
-  storage is more secure (due to encryption)
-  many users can write queries to gather insights from the data at the same time

**SQL**
- Structured Query Language
- most widely used programming language for creating, querying, and updating relational databases.
- answer questions both within and across relational database tables

**Tables**
- main building block of databases

In the world of databases:
- Rows are referred to as the **Records**, while Columns are referred to as the **Fields**

    **Fields**
    - are set at database creation
    - a column that holds one piece of information about all records
        
        _Generally, field names should:_
        
            - be lowercase
            - have no spaces
            - be singular (because it refers to the information contained in that field for a single record)
            - be different from other field names
            - be different from the table name
    
    **Records**
    - There is no limit to its number
    - a row that holds data on an individual observation

**Good Table Manners**

Table names should...
- be lowercase
- have no spaces-use underscores instead
- refer to a collective group or be plural (**patrons** is a good name, no to ~~Patrons~~, ~~library patrons~~, ~~patron~~)

**Unique Identifiers** _(Key)_
- used to identify records in the table
- the are unique and often numbers
![image-10](image-10.png)

**The more tables, the merrier**
- Having more tables, each with a clearly marked subject, is generally better than having fewer tables where information about multiple subjects is combined.

See photo below:
![image-11](image-11.png)
_(Table below is much less clear since it contains duplicate information)_

**Data**

**SQL Data Types**
- Different types of data are stored differently and take up different space
- Some operations only apply to certain data types

    **Strings**
    - a sequence of characters such as letters or punctuation
    - `VARCHAR`is a flexible, popular, and very commonly used string data type in SQL (because it can store small or large strings - up to tens of thousands of characters)
    
    **Integers**
    - store whole numbers
    - `INT` is a flexible and popular integer data type in SQL (because it can store numbers from less than negative 2 billion to more than positive two billion)

    **Float**
    - store numbers that include a fractional part
    - `NUMERIC`is a flexible and popular float data type in SQL (as it can store up to 38 digits total, including those before and after the decimal point).

**Schemas**
- often referred to as _"blueprints"_ of databases.
- shows the database's design, such as what tables are included in the database and any relationships between its tables.
- also lets the reader know what data type can each field hold.
![image-18](image-18.png)

**Database storage**
- _Servers_ - centralized computers that perform services (i.e., data access) via requests made over a network
- Servers are generally very powerful and large machines as they are equipped to handle large volume of requests and data.

In [2]:
-- Add your own queries here
SELECT * 
FROM books

id  ...        genre
0      1.0  ...  Non Fiction
1      2.0  ...      Fiction
2      3.0  ...  Non Fiction
3      4.0  ...      Fiction
4      5.0  ...  Non Fiction
..     ...  ...          ...
347  346.0  ...  Non Fiction
348  347.0  ...      Fiction
349  348.0  ...  Non Fiction
350  349.0  ...      Fiction
351  350.0  ...      Fiction

[352 rows x 5 columns]

### Chapter 2 - Querying

**What is SQL useful for?**
- Best for large datasets
- Used as a complement to other tools such as spreadsheet applications
- For sprawling _(expanding or extending in a disorderly or uncontrolled way)_ and diverse data, organizing the data in a database
- We use SQL queries to uncover trends in website traffic, customer reviews, and product sales.
- Used when we have lots of data with complex relationships

**Keywords**
- _Keywords_ are reserved words to indicate what operation we'd like our code to perform
- Common keywords: `SELECT`,`FROM`

`SELECT` keyword
- indicates which field should be selected

`FROM` keyword
- indicates the table in which these fields are located

 **Notes**
- `SELECT` statement appears first, followed by the `FROM` stament
- It is best a best practice to end the query with a semicolon (`;`) to indicate that the query is complete
- Keywords are capitalized while keeping table and field names all lowercase
- Naka-arrange yung field names according sa order na gusto natin ipalabas sa Result Set

**Result set**
- it is the query results



In [7]:
-- Our first query
SELECT title 
FROM books;

title                    author
0                        10-Day Green Smoothie Cleanse                  JJ Smith
1                                    11/22/63: A Novel              Stephen King
2              12 Rules for Life: An Antidote to Chaos        Jordan B. Peterson
3                               1984 (Signet Classics)             George Orwell
4    5,000 Awesome Facts (About Everything!) (Natio...  National Geographic Kids
..                                                 ...                       ...
347  Wild: From Lost to Found on the Pacific Crest ...            Cheryl Strayed
348  Winter of the World: Book Two of the Century T...               Ken Follett
349  Women Food and God: An Unexpected Path to Almo...               Geneen Roth
350                                             Wonder             R. J. Palacio
351       Wrecking Ball (Diary of a Wimpy Kid Book 14)               Jeff Kinney

[352 rows x 2 columns]

In [3]:
-- Selecting multiple fields
SELECT title, author -- just add comma
FROM books;

title                    author
0                        10-Day Green Smoothie Cleanse                  JJ Smith
1                                    11/22/63: A Novel              Stephen King
2              12 Rules for Life: An Antidote to Chaos        Jordan B. Peterson
3                               1984 (Signet Classics)             George Orwell
4    5,000 Awesome Facts (About Everything!) (Natio...  National Geographic Kids
..                                                 ...                       ...
347  Wild: From Lost to Found on the Pacific Crest ...            Cheryl Strayed
348  Winter of the World: Book Two of the Century T...               Ken Follett
349  Women Food and God: An Unexpected Path to Almo...               Geneen Roth
350                                             Wonder             R. J. Palacio
351       Wrecking Ball (Diary of a Wimpy Kid Book 14)               Jeff Kinney

[352 rows x 2 columns]

In [4]:
-- Selecting all fields
SELECT * -- use asterisk to select all fields
FROM books;

id  ...        genre
0      1.0  ...  Non Fiction
1      2.0  ...      Fiction
2      3.0  ...  Non Fiction
3      4.0  ...      Fiction
4      5.0  ...  Non Fiction
..     ...  ...          ...
347  346.0  ...  Non Fiction
348  347.0  ...      Fiction
349  348.0  ...  Non Fiction
350  349.0  ...      Fiction
351  350.0  ...      Fiction

[352 rows x 5 columns]

**Aliasing**

- Use aliasing to rename columns (just for clarity/brevity)
- it only applies to the result of a particular query result

In [10]:
SELECT year AS year_pub, id, title, author, genre
FROM books;

year_pub     id  ...                    author        genre
0      2016.0    1.0  ...                  JJ Smith  Non Fiction
1      2011.0    2.0  ...              Stephen King      Fiction
2      2018.0    3.0  ...        Jordan B. Peterson  Non Fiction
3      2017.0    4.0  ...             George Orwell      Fiction
4      2019.0    5.0  ...  National Geographic Kids  Non Fiction
..        ...    ...  ...                       ...          ...
347    2012.0  346.0  ...            Cheryl Strayed  Non Fiction
348    2012.0  347.0  ...               Ken Follett      Fiction
349    2010.0  348.0  ...               Geneen Roth  Non Fiction
350    2013.0  349.0  ...             R. J. Palacio      Fiction
351    2019.0  350.0  ...               Jeff Kinney      Fiction

[352 rows x 5 columns]

In [1]:
-- Selecting distinct records
SELECT DISTINCT year
FROM books;

year
0   2013.0
1      NaN
2   2015.0
3   2011.0
4   2010.0
5   2014.0
6   2017.0
7   2019.0
8   2016.0
9   2012.0
10  2009.0
11  2018.0

In [2]:
-- Selecting distinct record with multiple fields
SELECT DISTINCT year, author
FROM books

year               author
0    2015.0      Brandon Stanton
1    2013.0        Rush Limbaugh
2    2018.0           Celeste Ng
3    2019.0  The Washington Post
4    2009.0         Gary Chapman
..      ...                  ...
309  2011.0        Heidi Murkoff
310  2012.0           John Green
311  2018.0          Craig Smith
312  2013.0         Zhi Gang Sha
313  2009.0     Kathryn Stockett

[314 rows x 2 columns]

**Views**
- A view is a virtual table that is the result of a saved SQL `SELECT` statement
- When accessed, views automatically updates in response to updates in the underlying data
- No result set when creating a view

In [ ]:
CREATE VIEW author_genre_title AS
SELECT author, genre, title
FROM books;

Error: CREATE VIEW author_genre_title AS
SELECT author, genre, title
FROM books; - permission denied for schema public

In [ ]:
--- Using views
SELECT author, genre
FROM author_genre_title;

Error: SELECT author, genre
FROM author_genre_title; - relation "author_genre_title" does not exist

**SQL Flavors**
- Both free and paid
- All used with relational databases
- Vast majority of keywords are the same
- All SQL flavors must follow universal standards
- Only the additions on top of these standards make flavors different

**Two Popular SQL Flavors**

1. _PostgreSQL_
- Free and open-source relational database system
- Created at the University of California, Berkeley
- "PostgreSQL" refers to both the PostgreSQL database system and its associated SQL flavor

2. _SQL Server_
- Has free and paid versions
- Created by Microsoft
- T-SQL is Microsoft's SQL flavor, used with SQL Server database

**Comparing PostgreSQL and SQL Server**
- Like dialects of the same language
Example:

PostgreSQL:
    
        SELECT id, name
        FROM employees
        LIMIT 2;

SQL Server:

        SELECT TOP(2) id, name
        FROM employees;

# **Exercises**

__What are the advantages of databases?__

Imagine you are part of a discussion at work about whether or not to create a database. You've learned about several advantages of storing data in a database rather than other traditional formats like spreadsheets.

See if you can remember what they are by selecting all of the advantages.
![image-4](image-4.png)

__Data organization__

If you'd like to use SQL to gain insights from data, understanding the organization of a database is an important first step. Take a look at the database below. Which of the following statements correctly describes its organization?
![image-5](image-5.png)
![image-8](image-8.png)

__Picking a unique ID__

You've learned that a unique identifier is a unique value that identifies a record so that it can be distinguished from other records in the same table.

Let's take a closer look at the employees table. Which of the fields do you think is best suited to be a unique identifier?
![image-12](image-12.png)
![image-13](image-13.png)

__Setting the table in style__

Imagine that you are designing a database and the following table has been suggested. Your task is to provide feedback on how this table could be improved. Use the skills you learned in the last video to critique it!
![image-15](image-15.png)
![image-16](image-16.png)

__Our very own table__

We've set up a database inside this course, and the `books` table is available in the exercise. You'll use SQL to query this table in the next chapter, but for now, it's time to explore what data `books` holds!

Your task is to choose the option below that best describes the information contained in `books`. There's no need to do any coding in this exercise; you can answer this question by looking at the `books` table in the console next to the words "query result." Because some book titles are long, you may need to scroll to the right in order to see all the information that the `books` table contains.

![image-17](image-17.png)

__At your service

Now that you know more about how data is stored, it's time to test those skills!

Select the statement about database storage that is false.
![image-20](image-20.png)

__Finding data types__

Imagine that you are starting a new job and have just started getting to know your new employer's database. You know that it's important to know the data type—such as `VARCHAR`, `INT`, or `NUMERIC`—corresponding to each field in a table. Where could you find this information?

![image-21](image-21.png)

__Choice of type__

You've learned that when a table is created, a data type must be indicated for each field. Choosing the correct data type allows the data to be stored correctly and makes certain operations associated with that data type available. For example, mathematical operations can be performed on NUMERIC and INT data types, but not on VARCHAR data. Thus, it makes sense to store numerical values as NUMERIC or INT so that you can perform math operations on them if needed.

In this exercise, you'll practice selecting the proper data type for your data!

Drag the field description to the bucket indicating the best data type to use for that field.

![image-23](image-23.png)

__SQL strengths__

Which of the below scenarios describes a situation in which using SQL would be useful?
![image-24](image-24.png)

__Developing SQL style__

Recall from the video that it's important to pay attention to the formatting of SQL queries in order to make them readable. This is especially helpful as you learn more keywords and your queries get longer.

In this exercise, you'll review the below query about the patrons table. This code will run properly, but it is messy and hard to read. Your task is to determine which edits are appropriate to improve the query so that it follows best practices for SQL style.

        SELECT CARD_NUM, TOTAL_FINE 
        from patrons

Here's a reminder about what the patrons table looks like!
![image-25](image-25.png)

Drag each suggestion to the proper zone depending on whether or not it will improve the query's style.
![image-26](image-26.png)

__Querying the books table__

You're ready to practice writing your first SQL queries using the `SELECT` and `FROM` keywords. Recall from the video that `SELECT` is used to choose the fields that will be included in the result set, while `FROM` is used to pick the table in which the fields are listed.

Feel free to explore `books` in the exercise. Let's zoom in on this table in the database schema to see the fields and data types it contains.

Your task in this exercise is to practice selecting fields from `books.`

1. Use SQL to return a result set of all book titles included in the `books` table.
-- Return all titles from the books table

        SELECT id, title, author, genre, year
        FROM books;

2. Select both the `title` and `author` fields from `books.`
-- Select title and author from the books table

        SELECT title, author
        FROM books;

3. Select all fields from the `books` table.

-- Select all fields from the books table
       
        SELECT *
        FROM books;

**Making queries DISTINCT**

You've learned that the `DISTINCT` keyword can be used to return unique values in a field. In this exercise, you'll use this understanding to find out more about the `books` table!

There are 350 books in the `books` table, representing all of the books that our local library has available for checkout. But how many different authors are represented in these 350 books? The answer is surely less than 350. For example, J.K. Rowling wrote all seven Harry Potter books, so if our library has all Harry Potter books, seven books will be written by J.K Rowling. There are likely many more repeat authors!

1. Write SQL code that returns a result set with just one column listing the unique authors in the `books` table.

-- Select unique authors from the books table

        SELECT DISTINCT author
        FROM books;

2. Update the code to return the unique `author` and `genre` combinations in the `books` table.

-- Select unique authors and genre combinations from the books table

    SELECT DISTINCT author, genre
    FROM books;
    
**Aliasing**

While the default column names in a SQL result set come from the fields they are created from, you've learned that aliasing can be used to rename these result set columns. This can be helpful for clarifying the intent or contents of the column.

Your task in this exercise is to incorporate an alias into one of the SQL queries that you worked with in the previous exercise!

-- Alias author so that it becomes unique_author

    SELECT DISTINCT author AS unique_author
    FROM books;

**VIEWing your query**

You've worked hard to create the below SQL query:

    SELECT DISTINCT author AS unique_author
    FROM books;
    
What if you'd like to be able to refer to it later, or allow others to access and use the results? The best way to do this is by creating a view. Recall that a view is a virtual table: it's very similar to a real table, but rather than the data itself being stored, the query code is stored for later use.

1. Add a single line of code that saves the results of the written query as a view called `library_authors.`

-- Save the results of this query as a view called library_authors

    CREATE VIEW library_authors AS
    SELECT DISTINCT author AS unique_author
    FROM books;

2. Check that the view was created by selecting all columns from `library_authors`.

-- Your code to create the view:
    
    CREATE VIEW library_authors AS
    SELECT DISTINCT author AS unique_author
    FROM books;

-- Select all columns from library_authors

    SELECT *
    FROM library_authors;

**Comparing flavors**

The video introduced several differences between SQL Server and PostgreSQL. These are just two of many relational database systems available, but the differences are fairly representative of the types of differences you'll see between other SQL database systems and flavors as well.

Drag the statement to the database management system that it correctly describes.
![image-27](image-27.png)

**Limiting results**

Let's take a look at a few of the genres represented in our library's books.

Recall that limiting results is useful when testing code since result sets can have thousands of results! Queries are often written with a `LIMIT` of just a few records to test out code before selecting thousands of results from the database.

Let's practice with `LIMIT`!

Using PostgreSQL, select the `genre` field from the `books` table; limit the number of results to 10.

-- Select the first 10 genres from books using PostgreSQL

    SELECT genre
    FROM books
    LIMIT 10;

**Translating between flavors**

In the previous exercise, you wrote the following code using PostgreSQL:

    SELECT genre
    FROM books
    LIMIT 10;
    
The database in this course is a PostgreSQL database, so you won't be able to run SQL Server code in any of the exercises. What if you did want to update the above query to work with SQL Server, though? How would you do that?
![image-29](image-29.png)


# Congratulations!

- Learned when and why databases are useful
- Navigated the organization of relational databases
- Written own SQL queries to extract insight from a database

**Where to go next**
- Learn more keywords and choose which flavor you'll learn

![image-30](image-30.png)


## Explore Datasets
Use the `books` table to explore the data and practice your skills!
- Select only the `title` column.
- Alias the `title` column as `book_title`.
- Select the distinct author names from the `author` column.
- Select all records from the table and limit your results to 10.